test/uri/uri_parse_test.cc

```diff
@@ -0,0 +1,139 @@
+      sourcemeta::jsontoolkit::URIParseError);
+}
+
+// Inspired from
+// https://github.com/uriparser/uriparser/blob/bf0174e83164a4659c51c135399478bec389eafa/test/test.cpp#L315
+
+TEST(URI_parse, success_1) {
+  sourcemeta::jsontoolkit::URI uri{
+      "//user:pass@[::1]:80/segment/index.html?query#frag"};
+  EXPECT_EQ(uri.recompose(),
+            "user:pass@[::1]:80/segment/index.html?query#frag");
+}
+
+TEST(URI_parse, success_2) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://[::1]:80/segment/index.html?query#frag"};
+  EXPECT_EQ(uri.recompose(), "http://[::1]:80/segment/index.html?query#frag");
+}
+
+TEST(URI_parse, success_3) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://user:pass@[::1]/segment/index.html?query#frag"};
+  EXPECT_EQ(uri.recompose(),
+            "http://user:pass@[::1]/segment/index.html?query#frag");
+}
+
+TEST(URI_parse, success_4) {
+  sourcemeta::jsontoolkit::URI uri{"http://user:pass@[::1]:80?query#frag"};
+  EXPECT_EQ(uri.recompose(), "http://user:pass@[::1]:80?query#frag");
+}
+
+TEST(URI_parse, success_5) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://user:pass@[::1]:80/segment/index.html#frag"};
+  EXPECT_EQ(uri.recompose(),
+            "http://user:pass@[::1]:80/segment/index.html#frag");
+}
+
+TEST(URI_parse, success_6) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://user:pass@[::1]:80/segment/index.html?query"};
+  EXPECT_EQ(uri.recompose(),
+            "http://user:pass@[::1]:80/segment/index.html?query");
+}
+
+TEST(URI_parse, success_7) {
+  sourcemeta::jsontoolkit::URI uri{"ftp://host:21/gnu/"};
+  EXPECT_EQ(uri.recompose(), "ftp://host:21/gnu/");
+}
+
+TEST(URI_parse, relative_1) {
+  sourcemeta::jsontoolkit::URI uri{"one/two/three"};
+  EXPECT_EQ(uri.recompose(), "one/two/three");
+}
+
+TEST(URI_parse, relative_2) {
+  sourcemeta::jsontoolkit::URI uri{"/one/two/three"};
+  EXPECT_EQ(uri.recompose(), "one/two/three");
+}
+
+TEST(URI_parse, relative_3) {
+  sourcemeta::jsontoolkit::URI uri{"//user:pass@localhost/one/two/three"};
+  EXPECT_EQ(uri.recompose(), "user:pass@localhost/one/two/three");
+}
+
+TEST(URI_parse, real_life_1) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://sourceforge.net/projects/uriparser/"};
+  EXPECT_EQ(uri.recompose(), "http://sourceforge.net/projects/uriparser/");
+}
+
+TEST(URI_parse, real_life_2) {
+  sourcemeta::jsontoolkit::URI uri{
+      "http://sourceforge.net/project/platformdownload.php?group_id=182840"};
+  EXPECT_EQ(
+      uri.recompose(),
+      "http://sourceforge.net/project/platformdownload.php?group_id=182840");
+}
+
+TEST(URI_parse, mailto) {
+  sourcemeta::jsontoolkit::URI uri{"mailto:test@example.com"};
+  EXPECT_EQ(uri.recompose(), "mailto:test@example.com");
+}
+
+TEST(URI_parse, relative_4) {
+  sourcemeta::jsontoolkit::URI uri{"../../"};
+  EXPECT_EQ(uri.recompose(), "../../");
+}
+
+TEST(URI_parse, root_path) {
+  sourcemeta::jsontoolkit::URI uri{"/"};
+  EXPECT_EQ(uri.recompose(), "/");
+}
```

Comment:
Fro this case, I had no choice other than check the original data (this.data), the good thing is that it helped me to skip a whole parsing job :)

Reply:
Bad that we did not have benchmark before :/ we could have seen an improvement.

Reply:
Haha nice! Yeah, we need more benchmarks. Overall a better process for evaluating performance and making sure we don't regress